mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ExecutorStateManagerImpl.java

```diff
@@ -508,32 +506,78 @@ private Optional<Map<TaskExecutorID, TaskExecutorState>> findTaskExecutorsFor(Ta
+        Optional<TaskExecutorGroupKey> bestGroupBySizeName = findBestGroupBySizeNameMatch(requestedConstraints);
+
+        return bestGroupBySizeName.isPresent()
+            ? bestGroupBySizeName
+            : findBestGroupByFitnessCalculator(requestedConstraints);
+    }
+
+    /**
+     * Finds the best fit Task Executor Group by matching sizeNames.
+     *
+     * @param requestedConstraints The constraints for the scheduling request.
+     * @return An Optional of the best fit Task Executor Group Key based on sizeName and scheduling attributes matching.
+     *         If no suitable key is found, it returns an empty Optional.
+     */
+    private Optional<TaskExecutorGroupKey> findBestGroupBySizeNameMatch(SchedulingConstraints requestedConstraints) {
+        return executorsByGroup.keySet()
+            .stream()
+            // Filter to retain groups where sizeName is present
+            .filter(group -> group.getSizeName().isPresent())
+            // Filter to retain groups where the requested sizeName is also present
+            .filter(group -> requestedConstraints.getSizeName().isPresent())
+            // Filter to retain groups where sizeNames of group and requested constraints are equal
+            .filter(group -> group.getSizeName().get().equalsIgnoreCase(requestedConstraints.getSizeName().get()))
+            // Verify scheduling attribute constraints
+            .filter(taskExecutorGroupKey -> areSchedulingAttributeConstraintsSatisfied(requestedConstraints,
+                taskExecutorGroupKey.getSchedulingAttributes()))
+            .findFirst();
+    }
+
+    /**
+     * Finds the best fit Task Executor Group by using a fitness calculator on machine definitions.
+     *
+     * Groups that match requestedConstraints and have a fitness score greater than 0 are considered.
+     * Among these, the key with the highest score is returned.
+     *
+     * @param requestedConstraints The constraints for the scheduling request.
+     * @return An Optional of the best fit Task Executor Group Key according to a fitness calculator.
+     *         If no suitable key is found, it returns an empty Optional.
+     */
+    private Optional<TaskExecutorGroupKey> findBestGroupByFitnessCalculator(SchedulingConstraints requestedConstraints) {
         return executorsByGroup.keySet()
             .stream()
-            // filter in the keys that have satisfied allocation constraints
-            .filter(taskExecutorGroupKey -> areSchedulingAttributeConstraintsSatisfied(requestedConstraints, taskExecutorGroupKey.getSchedulingAttributes()))
-            // Map each TaskExecutorGroupKey to a Pair containing the key and its corresponding fitness score
+            // Filter out if both sizeName exist and are different (ie. small vs large)
+            .filter(taskExecutorGroupKey -> {
+                Optional<String> teGroupSizeName = taskExecutorGroupKey.getSizeName();
```

Comment:
do we still need to care size names on the fallback case?

Reply:
My approach is as follows: if the scheduling request includes a size name, we should exclude any TEs that have a set size that differs from the request. We should only consider TEs that either have no size set at all or ones that also have the same size as specified in the scheduling request. This holds true even when we fallback to using the machine definition for matching. We want to maintain the integrity of the size label match regardless of the matching strategy used. Does it make sense?